src/ripple/rpc/impl/Role.cpp

```diff
@@ -122,18 +122,125 @@ requestInboundEndpoint(
+
+            ret = ret.substr(firstNonSpace);
+        }
+        // Trim trailing spaces.
+        if (!ret.empty())
+        {
+            // Only do the work if there's at least one trailing space.
+            if (unsigned char const c = ret.back();
+                c == ' ' || c == '\r' || c == '\n')
+            {
+                std::size_t const lastNonSpace = ret.find_last_not_of(" \r\n");
+                if (lastNonSpace == boost::string_view::npos)
+                    // We know there's at least one leading space.  So if we
+                    // got npos, then it must be all spaces.
+                    return {};
+
+                ret = ret.substr(0, lastNonSpace + 1);
+            }
+        }
+        return ret;
+    };
+
+    boost::string_view ret = trim(field);
+    if (ret.empty())
+        return {};
+
+    // If there are surrounding quotes, strip them.
+    if (ret.front() == '"')
+    {
+        ret.remove_prefix(1);
+        if (ret.empty() || ret.back() != '"')
+            return {};  // Unbalanced double quotes.
+
+        ret.remove_suffix(1);
+
+        // Strip leading and trailing spaces that were inside the quotes.
+        ret = trim(ret);
+    }
+    if (ret.empty())
+        return {};
+
+    // If we have an IPv6 or IPv6 (dual) address wrapped in square brackets,
+    // then we need to remove the square brackets.
+    if (ret.front() == '[')
+    {
+        // Remove leading '['.
+        ret.remove_prefix(1);
+
+        // We may have an IPv6 address in square brackets.  Scan up to the
+        // closing square bracket.
+        auto const closeBracket =
+            std::find_if_not(ret.begin(), ret.end(), [](unsigned char c) {
+                return std::isxdigit(c) || c == ':' || c == '.' || c == ' ';
+            });
+
+        // If the string does not close with a ']', then it's not valid IPv6
+        // or IPv6 (dual).
+        if (closeBracket == ret.end() || (*closeBracket) != ']')
+            return {};
+
+        // Remove trailing ']'
+        ret = ret.substr(0, closeBracket - ret.begin());
+        ret = trim(ret);
+    }
+    if (ret.empty())
+        return {};
+
+    // If this is an IPv6 address (after unwrapping from square brackets),
+    // then there cannot be an appended port.  In that case we're done.
+    {
+        // Skip any leading hex digits.
+        auto const colon =
+            std::find_if_not(ret.begin(), ret.end(), [](unsigned char c) {
+                return std::isxdigit(c) || c == ' ';
+            });
+
+        // If the string starts with optional hex digits followed by a colon
+        // it's an IVv6 address.  We're done.
+        if (colon == ret.end() || (*colon) == ':')
+            return ret;
+    }
+
+    // If there's a port appended to the IP address, strip that by
+    // terminating at the colon.
+    if (std::size_t colon = ret.find(':'); colon != boost::string_view::npos)
+        ret = ret.substr(0, colon);
+
+    return ret;
+}
+
```

Comment:
Would boost::regex be better? It's certainly a bit less code.

Reply:
This was a good question to ask.  At this prompt I explored using `boost::regex` to do the work.  I don't have a lot of experience with `boost::regex`, so maybe I'm missing something.  But from what I can see `boost::regex` does not support `boost::string_view` very well if at all.

The problem is that `ripple::forwardedFor()` is designed to return a `boost::string_view` extracted from the passed in `http_request_type const&`.  I believe this design is intended to avoid the memory allocations that are involved with `std::string` creation.  After poking at `boost::regex` for a few hours I've been unable to locate the operations I would need to perform in order to get `boost::regex` to operate on and produce the needed `boost::string_view`.  A few web searches helped to reinforce this suspicion.

Maybe you have more experience with `boost::regex` and can steer me in the right direction?  Failing that, I'm afraid the current character manipulations are the best I can do.

I did not look into using `std::regex` since, disappointingly, `std::regex` is not portable across platforms.

Reply:
I see. I didn't consider `boost::string_view`.